clients/src/main/java/org/apache/kafka/clients/consumer/internals/AbstractStickyAssignor.java

```diff
@@ -205,6 +237,9 @@ private boolean allSubscriptionsEqual(Set<String> allTopics,
                 // consumer owned the "maxQuota" of partitions or more, and we're still under the number of expected members
                 // with more than the minQuota partitions, so keep "maxQuota" of the owned partitions, and revoke the rest of the partitions
                 numMembersAssignedOverMinQuota++;
+                if (numMembersAssignedOverMinQuota == expectedNumMembersAssignedOverMinQuota) {
+                    potentiallyUnfilledMembersAtMinQuota.clear();
```

Comment:
I'd suggest we remove this `potentiallyUnfilledMembersAtMinQuota.clear();` logic and just keep two expected numbers:

1) expectedNumMembersWithMaxQuota = totalPartitionsCount % numberOfConsumers;
2) expectedNumMembersWithMinQuota = numberOfConsumers - expectedNumMembersWithMaxQuota;

And then we can also remove the check in line 309, and at the end when we exhausted `unassignedPartitions`, just check

```
numMembersWithMaxQuota == expectedNumMembersWithMaxQuota &&
numMembersWithMinQuota == expectedNumMembersWithMinQuota
```

WDYT?

Reply:
While I'm not really a fan of the `potentiallyUnfilledMembersAtMinQuota` logic (it's definitely awkward but I felt it was still the lesser evil in terms of complicating the code), I don't think we can get rid of it that easily. The problem is that when `minQuota != maxQuota`, and so far `currentNumMembersWithOverMinQuotaPartitions` < `expectedNumMembersWithOverMinQuotaPartitions`, then consumers that are filled up to exactly `minQuota` have to be considered potentially not yet at capacity since some will need one more partition, though not all. So this data structure is not just used to verify that everything is properly assigned after we've exhausted the `unassignedPartitions`, it's used to track which consumers can still receive another partition (ie, are "unfilled"). Does that make sense?

Reply:
Yes that makes sense, still this logic

```
if (numMembersAssignedOverMinQuota == expectedNumMembersAssignedOverMinQuota) {
                    potentiallyUnfilledMembersAtMinQuota.clear();
}
```

Seems only needed because we have the check in 309 (?) Say if we do not check that, but instead just check the expected numbers of consumers with minQuota and maxQuota is satisfied, then do we still need this?

Reply:
> this logic Seems only needed because we have the check in 309 (?)

No, I don't think so. It should be for line 279:
```java
// to handle the case that when there are still unassignedPartition left, but no more members to be assigned.
if (unfilledMembersWithUnderMinQuotaPartitions.isEmpty() && unfilledMembersWithExactlyMinQuotaPartitions.isEmpty()) {
   throw new IllegalStateException("No more unfilled consumers to be assigned.");
```

In line 309, it is just an early error detect and log for it. Not related to `potentiallyUnfilledMembersAtMinQuota` (or now `unfilledMembersWithExactlyMinQuotaPartitions` members)

Reply:
I tried to summarize both methods:

What @guozhangwang 's meaning is, we can "lazily" detect the issue after assigning all unassignedPartitions. we don't need to clear the `potentiallyUnfilledMembersAtMinQuota` here, because as the "original" variable naming said: they are "potentially unfilled members", just keep them there. We "should not" assign partitions to them in this case because we've reached `expectedNumMembersAssignedOverMinQuota`. 

But if somehow, after assigning unassignedPartitions to all unfilledMembers, there are still some unassignedPartitions left. We can just assign them to `potentiallyUnfilledMembersAtMinQuota`. And after running out the `unassignedPartitions`, we can check:
```
numMembersWithMaxQuota == expectedNumMembersWithMaxQuota &&
numMembersWithMinQuota == expectedNumMembersWithMinQuota
```

to do error handling. 

VS.

In @ableegoldman 's version , we find issue immediately and handle it. we computed the `potentiallyUnfilledMembersAtMinQuota` correctly (that's why we need to clear it). So, if the issue happened:  
> if somehow, after assigning unassignedPartitions to all unfilledMembers, there are still some unassignedPartitions left

We can try to get member from `potentiallyUnfilledMembersAtMinQuota` and then assign unassignedPartition to the member. If we can't get member from it (i.e. `potentiallyUnfilledMembersAtMinQuota` is empty), we throw exception directly.


Both ways can find errors when happened. Personally, I like Sophie's version more since it's much clear.

Reply:
Thanks @showuon and @guozhangwang , I think that all makes sense. One of my primary motivations was to keep all data structures at all times consistent with what they represent so they could always be relied upon to be used at any point. For that reason I also prefer to keep things as is, and clear the `potentiallyUnfilledMembersAtMinQuota` (now renamed to `nfilledMembersWithExactlyMinQuotaPartitions`) as soon as we have filled the last member who may be above `minQuota`, at which point all of the members at exactly `minQuota` are no longer considered "unfilled"

Reply:
SGTM.